Initial

gsteel opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | not yet

### Description

Initial pull request with filters, validators and view helpers.

gsteel committed:
Initial commit

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Initial project setup

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Update .gitattributes and rename phpcs file

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Laminas StdLib is a requirement for Array Utils

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Initial rev of PhoneNumber value object and the validator

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Cover traversable options

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Update to laminas coding standard 2.3

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Adds RFC3966 formatting to the phone number value object

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Add filters for formatting phone numbers

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Initial config provider with factories and configuration for filters

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Adds a factory for the validator

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Clean up composer.json adding the config provider.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Adds an alias for the validator that would in replace the one shipped with i18n

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Adds a method to format the phone number to an out of country calling number

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Initial rev of a view helper to conveniently format numbers

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Adds a factory for the view helper and provides config for the helper plugin manager

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel commented on src/PhoneNumberValue.php:
```diff
@@ -0,0 +1,224 @@
+    public const TYPE_RECOMMENDED = self::TYPE_FIXED
+                                  | self::TYPE_MOBILE
+                                  | self::TYPE_VOIP;
+
```
`TYPE_RECOMMENDED` is intended as a default value that is typical for what the "most" consumers are likely to accept as input.

gsteel commented on src/Filter/AbstractFilter.php:
```diff
@@ -0,0 +1,60 @@
+use function sprintf;
+use function strtoupper;
+
+abstract class AbstractFilter implements FilterInterface
```
Should be marked as internal

gsteel commented on src/ConfigProvider.php:
```diff
@@ -0,0 +1,81 @@
+        return [
+            'laminas-i18n-phone-number' => [
+                'default-country-code'    => null,
+                'acceptable-number-types' => PhoneNumberValue::TYPE_RECOMMENDED,
```
This is the only use of `TYPE_RECOMMENDED`, theoretically making it easy to override what is considered an acceptable number type in a place that is easy to document

gsteel commented on src/Validator/PhoneNumber.php:
```diff
@@ -0,0 +1,135 @@
+     *
+     * Restrict otherwise valid types of phone numbers to a subset of allowed types.
+     */
+    private int $allowTypes = PhoneNumberValue::TYPE_ANY;
```
Default behaviour for the validator is to accept any type of phone number

gsteel commented on src/PhoneNumberValue.php:
```diff
@@ -0,0 +1,224 @@
+
+use function array_key_exists;
+
+final class PhoneNumberValue
```
Should this value object be marked as internal?

Ocramius replied:
If you are not sure about this API being `public`, then yes, otherwise you will have a BC problem after `1.0.0`

gsteel replied:
I'm undecided about this - it's a useful tool outside of just validation - also, we don't want psalm disallowing the use of the constants for consumers because they'll be used to construct masks of acceptable types.

I don't like `__toString()` because it contains formatting logic. Perhaps it would be better to store the input string and use that as the `__toString` value. Thoughts?

internalsystemerror replied:
Shame we can't use enums for these type constants, sounds like a perfect use case.

gsteel replied:
As bit masks, they're more flexible - for example, when consumers configure what they accept as a valid phone number, with:
```php
'laminas-i18n-phone-number' => [
    'default-country-code'    => 'US',
    'acceptable-number-types' => PhoneNumberValue::TYPE_TOLL_FREE | PhoneNumberValue::TYPE_EMEGENCY,
],
```

gsteel commented on src/View/Helper/PhoneNumberFormat.php:
```diff
@@ -0,0 +1,120 @@
+     * @param non-empty-string      $number
+     * @param non-empty-string|null $country
+     */
+    public function toPhoneNumber(string $number, ?string $country = null): ?PhoneNumberValue
```
If we make the value object internal, this method can be made private

gsteel replied:
I'm going to make this private anyhow… It's weird fetching the value object inside the view - the formatting methods should solve most use-cases I can think of…

gsteel commented on test/NumberGeneratorTrait.php:
```diff
@@ -0,0 +1,120 @@
+/**
+ * @internal \LaminasTest\I18n
+ */
+trait NumberGeneratorTrait
```
`libphonenumber`'s apparently valid example numbers don't all validate. These data providers work-around some of the unexpected numbers that get returned - typically, we're filtering out tens of results from many hundreds-ish.

Fahl-Design commented on src/PhoneNumberValue.php:
```diff
@@ -0,0 +1,224 @@
+        if (! $dialingCode) {
+            return $givenCode;
+        }
+
```
@gsteel hm maybe this should throw an exception if the `$regionCode !== $givenCode` because

> `+44` is not from `US` 

I like this lib and will use it when it's ready :+1:

gsteel replied:
Thanks for reviewing @Fahl-Design!

The purpose of this method is to safely coalesce or detect the country code to a non null value if possible and is used only by `fromString($number, $maybeCountryCode)`.

The idea is:
```php
PhoneNumberValue::fromString('01234567890', 'GB'); // Valid UK, local number
PhoneNumberValue::fromString('+44 1234567890', 'CH'); // Valid UK, country argument ignored
PhoneNumberValue::fromString('01234567890', 'US'); // Exception, UK local number isn't a valid US number (I don't think)
PhoneNumberValue::fromString('01234567890', null); // Exception, local number, no idea which country
```

So returning the given code here when otherwise the code cannot be detected from the number is intentional - given code can still be null at this point.

It might be a good idea to rename the method though to better describe its purpose

froschdesign commented on composer.json:
```diff
@@ -0,0 +1,72 @@
+{
+    "name": "laminas/laminas-i18n-phone-number",
+    "description": "Validate and format internation phone numbers",
```
```suggestion
    "description": "Validate and format international phone numbers",
```

froschdesign commented on README.md:
```diff
@@ -1 +1,22 @@
+
+[![Build Status](https://github.com/laminas/laminas-i18n-phone-number/workflows/Continuous%20Integration/badge.svg)](https://github.com/laminas/laminas-i18n-phone-number/actions?query=workflow%3A"Continuous+Integration")
+
+This library provides …
```
```suggestion
Validate and format international phone numbers
```

gsteel replied:
Thanks @froschdesign :)

gsteel committed:
Fixes typo in composer description. Adds tag line to the readme. Updates locked dependencies

Signed-off-by: George Steel <george@net-glue.co.uk>

froschdesign commented on src/View/Helper/PhoneNumberFormat.php:
```diff
@@ -0,0 +1,120 @@
+use function preg_match;
+use function strtoupper;
+
+final class PhoneNumberFormat
```
@gsteel 
Maybe a topic for laminas-view 3: a view helper as a generic filter that allows all registered filters to be used with a chain. No need for extra formatter view helpers.

gsteel replied:
Possibly? Would you define an 'AbstractFilterHelper' and then encourage a configuration based approach for individual use cases? Or, a bit like exposing a plugin manager in the view so `$this->filterHelper()->{$filterAlias}($value)` - The latter doesn't allow for chains without introducing loads of state and requires magic, but the former would be trivial to write from scratch for specific use cases and would offer better type safety.

In this case, the filters _and_ the view helper are just conveniences that proxy to methods in the phone number value object.

This conversation has me wondering whether a form element(s) should be introduced in this package too?

froschdesign replied:
I had the following in mind:

```php
echo $this->filter('  <i>example</i>  ')->stripTags()->trim()->stringToUpper();
```
…or something similar.

> This conversation has me wondering whether a form element(s) should be introduced in this package too?

Good idea! 👍

gsteel replied:
https://github.com/laminas/laminas-view/pull/164

Hanmac commented:
Question: just for asking, why not using https://github.com/giggsey/libphonenumber-for-php as dependency or is there a difference i the data?

froschdesign commented:
@Hanmac 
> Question: just for asking, why not using https://github.com/giggsey/libphonenumber-for-php as dependency or is there a difference i the data?

See:

https://github.com/laminas/laminas-i18n-phone-number/blob/9466fd47c2a2c6a545a95bbe22aa57ba6624d299/composer.json#L37

https://github.com/laminas/laminas-i18n-phone-number/pull/1/files#diff-d2ab9925cad7eac58e0ff4cc0d251a937ecf49e4b6bf57f8b95aab76648a9d34R37

Hanmac commented:
@froschdesign ahh sorry i somehow missed that.

if i understand that right, this Project is an extraction of this https://github.com/laminas/laminas-i18n/blob/2.16.x/src/Validator/PhoneNumber.php class, right?

froschdesign commented:
@Hanmac 
> …extraction of this https://github.com/laminas/laminas-i18n/blob/2.16.x/src/Validator/PhoneNumber.php class…

Extraction and rework because the current data for the phone numbers are outdated and hard to update. `giggsey/libphonenumber-for-php` is great a library and provides a future-proof basis for laminas-i18n.

This component will contain filters and a view helper in addition to the validator.

froschdesign commented on README.md:
```diff
@@ -1 +1,22 @@
+
+## Documentation
+
+Browse the documentation online at https://docs.laminas.dev/laminas-i18n-phone-number/
```
```suggestion
Browse the documentation online at https://docs.laminas.dev/laminas-i18n/
```

froschdesign commented on composer.json:
```diff
@@ -0,0 +1,72 @@
+        "validation",
+        "phonenumber"
+    ],
+    "homepage": "https://laminas.dev",
```
```suggestion
    "homepage": "https://getlaminas.org",
```

froschdesign commented on composer.json:
```diff
@@ -0,0 +1,72 @@
+    ],
+    "homepage": "https://laminas.dev",
+    "support": {
+        "docs": "https://docs.laminas.dev/laminas-i18n-phone-number/",
```
```suggestion
        "docs": "https://docs.laminas.dev/laminas-i18n/",
```

froschdesign commented on composer.json:
```diff
@@ -0,0 +1,72 @@
+    },
+    "extra": {
+        "laminas": {
+            "config-provider": "Laminas\\I18n\\PhoneNumber\\ConfigProvider"
```
This component should also work in a laminas-mvc based application.

froschdesign commented on .github/workflows/docs-build.yml:
```diff
@@ -0,0 +1,16 @@
+      - name: Build Docs
+        uses: laminas/documentation-theme/github-actions/docs@master
+        env:
+          DOCS_DEPLOY_KEY: ${{ secrets.DOCS_DEPLOY_KEY }}
```
The documentation of this component will be integrated in laminas-i18n, therefore this can be omitted.

froschdesign commented on docs/book/index.md:
```diff
@@ -0,0 +1 @@
+../../README.md
```
Please ignore the documentation completely in this pull request, therefore remove this file.

froschdesign commented on docs/book/v1/intro.md:
```diff
@@ -0,0 +1,3 @@
+# laminas-i18n-phone-number
```
Please remove this file.

froschdesign commented on mkdocs.yml:
```diff
@@ -0,0 +1,13 @@
+docs_dir: docs/book
```
Please remove this file.

froschdesign commented on src/View/Helper/PhoneNumberFormat.php:
```diff
@@ -0,0 +1,120 @@
+     * @param non-empty-string|null $defaultCountryCode
+     * @throws InvalidArgumentException If the country code is not a valid ISO 3166 string.
+     */
+    public function __construct(?string $defaultCountryCode = null)
+    {
+        $this->country = self::assertValidCountryCode($defaultCountryCode);
+    }
```
With this new handling of the country code we have to deprecate the current setting and handling of a locale in all existing view helpers of laminas-view.

Example: https://docs.laminas.dev/laminas-i18n/view-helpers/number-format/#using-locale

Otherwise we create an inconsistency.

gsteel replied:
This has got me thinking that a CountryCode value object would be more useful instead of dragging around and validating the string value all the time… Something like:

```php
final class CountryCode
{
  private function __construct(public readonly string $code)
  {
  }

  public static function fromLocaleString(string $locale): self;

  public static function fromString(string $code): self;

  public static function fromLocale(Locale $locale): self;

  public static function fromMixed(mixed $value): self;
}
```

This would be more flexible and provide better compat with i18n right? i.e. the view helper constructor would then become…

```php
public function __construct(?mixed $localeOrCountryCode = null)
{
  $localeOrCountryCode = $localeOrCountryCode ?: Locale::getDefault();
  $this->countryCode = CountryCode::fromMixed($localeOrCountryCode);
}
```

WDYT?

froschdesign replied:
> …a CountryCode value object would be more useful instead of dragging around and validating the string value all the time…

The value object is a good idea and the correct way to go. 👍

froschdesign reviewed (changes requested):
Very good start! 👍 ❤️ 

---

All files and parts for the documentation should be removed from this pull request because the documentation will be integrated in laminas-i18n. We will add everything that is necessary later.

gsteel committed:
Remove documentation assets

- Documentation for this package will be part of laminas-i18n so links are changed and docs related files removed

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Adds a Module class for MVC applications

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Introduce country code value object

This value object should be used anywhere a country code string is expected

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Modify filters, validator and view helper to accept a locale string or a country code

The country code becomes a required constructor parameter for the filters and view helper because it now falls back to the default locale in the relevant factories.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel commented:
Thanks for the review @froschdesign - Could you please look over the introduction of the CountryCode value object and how this falls back to `Locale::getDefault()`

Also, could you take a look at the settings to see if you can get CI running?

Paging @Ocramius for review too - I'd like to get something into a 1.0.x branch so this pull doesn't get too ugly

froschdesign commented on src/CountryCode.php:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    /** @param non-empty-string $locale */
+    public static function fromLocale(string $locale): self
```
I am a little confused with the method names because `fromString()` creates a the value object from a string. The name of the method `fromLocale()` also creates the object from a string, but not from the PHP's Locale `Locale` as the name of the method suggests.
The method signatures are clear, but the name and their handling confuse me a little.

gsteel replied:
Thanks @froschdesign - I see your point. Perhaps `fromString` should be renamed to `fromCode`? PHP's Locale objects can't be instantiated - it's just a bunch of static methods.

Would changing the `fromString` method name remove potential confusion do you think?

marcguyer replied:
IMO, `fromString()` is fine, but I think @froschdesign means `fromLocale()` should become `fromLocaleString()` for clarity.

gsteel committed:
rename `fromLocale` to `fromLocaleString` fore less ambiguity

Signed-off-by: George Steel <george@net-glue.co.uk>

Ocramius commented on COPYRIGHT.md:
```diff
@@ -0,0 +1 @@
+Copyright (c) 2022 Laminas Project a Series of LF Projects, LLC. (https://getlaminas.org/)
```
I suggest we drop the year <_<

Ocramius commented on LICENSE.md:
```diff
@@ -0,0 +1,26 @@
+Copyright (c) 2022 Laminas Project a Series of LF Projects, LLC.
```
I suggest we drop the year

Ocramius commented on composer.json:
```diff
@@ -0,0 +1,74 @@
+        }
+    },
+    "require": {
+        "php": "^7.4 || ~8.0.0 || ~8.1.0",
```
Non-blocker, but do we need 7.4 on a new package? :thinking: 

This one could really use all the new `readonly` goodness from 8.1 too

gsteel replied:
I'm good with that :) 7.4 and 8.0 are dropped - I'll go through the rest of it making use of 8.1 where possible.

Ocramius commented on composer.json:
```diff
@@ -0,0 +1,74 @@
+    },
+    "autoload": {
+        "psr-4": {
+            "Laminas\\I18n\\PhoneNumber\\": "src//"
```
Double `/` here

Ocramius commented on psalm-baseline.xml:
```diff
@@ -0,0 +1,2 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<files psalm-version="4.21.0@d8bec4c7aaee111a532daec32fb09de5687053d1"/>
```
Do we need a baseline at all? WDYT?

Also worth adding a comment on why this baseline is currently empty, and perhaps when/if we expect it to expand in future

gsteel replied:
I think it's worth leaving so it's not deleted/added over time (??) Added a note…

Ocramius commented on src/ConfigProvider.php:
```diff
@@ -0,0 +1,81 @@
+ */
+final class ConfigProvider
+{
+    public function __invoke(): array
```
Should we document the return value in a more precise way, or is that just wasteful?

gsteel replied:
Done it… Same for `Module`

Ocramius commented on src/CountryCode.php:
```diff
@@ -0,0 +1,115 @@
+final class CountryCode
+{
+    /**
+     * @readonly
```
Not necessary with `@psalm-immutable` on the class

Ocramius commented on src/CountryCode.php:
```diff
@@ -0,0 +1,115 @@
+        return $this->code === $other->code;
+    }
+
+    /** @param non-empty-string $code */
```
`@psalm-pure`?

Ocramius commented on src/CountryCode.php:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    /** @param non-empty-string $code */
+    public static function fromString(string $code): self
```
We should document the `$code` parameter in `@param`, providing examples, perhaps.

Ocramius commented on src/CountryCode.php:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    /** @param non-empty-string $locale */
+    public static function fromLocaleString(string $locale): self
```
We should document/describe `$locale` in the docblock

Ocramius commented on src/CountryCode.php:
```diff
@@ -0,0 +1,115 @@
+        return new self($code);
+    }
+
+    /** @param non-empty-string $locale */
```
`@psalm-pure` too?

Ocramius commented on src/CountryCode.php:
```diff
@@ -0,0 +1,115 @@
+        ));
+    }
+
+    /** @param non-empty-string $countryCodeOrLocale */
```
`@psalm-pure`?

Ocramius commented on src/CountryCode.php:
```diff
@@ -0,0 +1,115 @@
+        }
+
+        if ($countryCodeOrLocale === null || $countryCodeOrLocale === '') {
+            $countryCodeOrLocale = Locale::getDefault();
```
Note: `@psalm-pure` not possible here, due to to this specific call

Ocramius commented on src/CountryCode.php:
```diff
@@ -0,0 +1,115 @@
+
+        $displayName = Locale::getDisplayRegion('-' . $code, 'GB');
+        if ($displayName === '' || $displayName === 'Unknown Region') {
+            throw new InvalidArgumentException(sprintf(
+                'The country code "%s" does not correspond to a known country',
+                $code
+            ));
```
Let's use a named ctor on this exception perhaps?

Ocramius commented on src/CountryCode.php:
```diff
@@ -0,0 +1,115 @@
+        $region = Locale::getRegion($locale);
+        /** @psalm-suppress TypeDoesNotContainNull */
+        if ($region === null || $region === '') {
+            throw new InvalidArgumentException(sprintf(
+                'The string "%s" could not be parsed as a valid locale',
+                $locale
+            ));
```
Let's use a named ctor on this exception perhaps?

Ocramius commented on src/CountryCode.php:
```diff
@@ -0,0 +1,115 @@
+            return $code;
+        }
+
+        throw new InvalidArgumentException(sprintf(
+            'The string "%s" could not be understood as either a locale or an ISO 3166 country code',
+            $countryCodeOrLocale
+        ));
```
Let's use a named ctor on this exception perhaps?

Ocramius commented on src/Exception/InvalidPhoneNumberException.php:
```diff
@@ -0,0 +1,41 @@
+{
+    public static function with(string $phoneNumber, ?string $countryCode): self
+    {
+        if (! empty($countryCode)) {
```
Let's not use `empty()`, please (like ever :D )

`empty($v)` is equivalent to `! isset($v) || $v === '' || $v === false || $v === 0 || $v === '0' || $v === [] || $v === 0.0` (and more stuff perhaps - can't even remember).

Never use it, if possible.

Ocramius commented on src/Exception/InvalidPhoneNumberException.php:
```diff
@@ -0,0 +1,41 @@
+
+final class InvalidPhoneNumberException extends InvalidArgumentException implements ExceptionInterface
+{
+    public static function with(string $phoneNumber, ?string $countryCode): self
```
`@psalm-pure`?

Ocramius commented on src/Exception/InvalidPhoneNumberException.php:
```diff
@@ -0,0 +1,41 @@
+        ));
+    }
+
+    public static function invalidNumberForRegion(string $phoneNumber, string $countryCode): self
```
`@psalm-pure`?

Ocramius commented on src/Exception/InvalidPhoneNumberException.php:
```diff
@@ -0,0 +1,41 @@
+        ));
+    }
+
+    public static function undetectableRegion(string $phoneNumber): self
```
`@psalm-pure`?

Ocramius commented on src/Exception/UnrecognizableNumberException.php:
```diff
@@ -0,0 +1,21 @@
+
+final class UnrecognizableNumberException extends InvalidArgumentException implements ExceptionInterface
+{
+    public static function withString(string $number, ?Throwable $previous = null): self
```
`@psalm-pure`?

Ocramius commented on src/Factory/ConfigurationTrait.php:
```diff
@@ -0,0 +1,41 @@
+ * @psalm-import-type PackageConfig from ConfigProvider
+ * @psalm-internal \Laminas\I18n\PhoneNumber
+ */
+trait ConfigurationTrait
+{
+    private function defaultCountryCode(ContainerInterface $container): CountryCode
+    {
+        $options = $this->componentConfig($container);
+
+        return CountryCode::detect($options['default-country-code'] ?? null);
+    }
+
+    /** @return PackageConfig */
+    private function componentConfig(ContainerInterface $container): array
+    {
+        if (! $container->has('config')) {
+            return [];
+        }
+
+        /** @psalm-var array $config */
+        $config = $container->get('config');
+
+        /** @psalm-var PackageConfig|null $options */
+        $options = $config['laminas-i18n-phone-number'] ?? null;
+
+        return is_array($options) ? $options : [];
+    }
```
I would make this into an `abstract class`, then declare both methods `public static` and use them directly from other classes.

No need to introduce the complexity of a `trait`

Ocramius commented on src/Factory/ConfigurationTrait.php:
```diff
@@ -0,0 +1,41 @@
+        /** @psalm-var PackageConfig|null $options */
+        $options = $config['laminas-i18n-phone-number'] ?? null;
+
+        return is_array($options) ? $options : [];
```
Could be `$options ?? []`?

Ocramius commented on src/Filter/AbstractFilter.php:
```diff
@@ -0,0 +1,45 @@
+            return null;
+        }
+
+        $input = (string) $value;
```
What if it is an `object` that cannot be cast to `string`? :thinking:

gsteel replied:
It's already been checked as a scalar

Ocramius commented on src/Filter/AbstractFilter.php:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    /** @param mixed $value */
+    protected function mixedToPhoneNumber($value): ?PhoneNumberValue
```
I think this is `tryMixedToPhoneNumber`

Ocramius commented on src/Filter/Factory/ToE164Factory.php:
```diff
@@ -0,0 +1,19 @@
+
+final class ToE164Factory
+{
+    use ConfigurationTrait;
```
Discussed above: let's avoid using the trait, and rather use static methods instead

Ocramius commented on src/Filter/Factory/ToInternationalPhoneNumberFactory.php:
```diff
@@ -0,0 +1,19 @@
+
+final class ToInternationalPhoneNumberFactory
+{
+    use ConfigurationTrait;
```
Discussed above: let's avoid using the trait, and rather use static methods instead

Ocramius commented on src/Filter/Factory/ToNationalPhoneNumberFactory.php:
```diff
@@ -0,0 +1,19 @@
+
+final class ToNationalPhoneNumberFactory
+{
+    use ConfigurationTrait;
```
Discussed above: let's avoid using the trait, and rather use static methods instead

Ocramius commented on src/Filter/ToE164.php:
```diff
@@ -0,0 +1,24 @@
+     * @param T $value
+     * @return string|T
+     */
+    public function filter($value)
+    {
+        $number = $this->mixedToPhoneNumber($value);
+
+        return $number ? $number->toE164() : $value;
```
Does this filter actually filter? Should it really ignore non-phone-numbers? :thinking:

gsteel replied:
The way most of the laminas filters are implemented, appear to me, to always attempt to filter a mixed value if possible or return the input if not - exceptions are only thrown when there's a configuration problem. I'm sure there are exceptions to the rule but I just scanned through a bunch of standard filters and they all seem to operate this way.

Ocramius commented on src/Filter/ToInternationalPhoneNumber.php:
```diff
@@ -0,0 +1,24 @@
+     * @param T $value
+     * @return string|T
+     */
+    public function filter($value)
+    {
+        $number = $this->mixedToPhoneNumber($value);
+
+        return $number ? $number->toInternational() : $value;
```
Does this filter actually filter? Should it really ignore non-phone-numbers? :thinking:

Ocramius commented on src/Filter/ToNationalPhoneNumber.php:
```diff
@@ -0,0 +1,24 @@
+     * @param T $value
+     * @return string|T
+     */
+    public function filter($value)
+    {
+        $number = $this->mixedToPhoneNumber($value);
+
+        return $number ? $number->toNational() : $value;
```
Does this filter actually filter? Should it really ignore non-phone-numbers? :thinking:

Ocramius commented on src/Module.php:
```diff
@@ -0,0 +1,13 @@
+
+namespace Laminas\I18n\PhoneNumber;
+
+final class Module
```
We could drop a line here, explaining that this is for compatibility with the module manager?

Ocramius commented on src/ConfigProvider.php:
```diff
@@ -0,0 +1,81 @@
+ *     default-country-code?: non-empty-string|null
+ * }
+ */
+final class ConfigProvider
```
We could drop a line here, explaining that this is for usage in some sort of config provider?

Ocramius commented on src/PhoneNumberValue.php:
```diff
@@ -0,0 +1,224 @@
+            ? $regionCode
+            : $givenCode;
+
+        return empty($regionCode) ? null : $regionCode;
```
Similar discussion as above around `empty()`

Ocramius commented on src/Validator/Factory/PhoneNumberFactory.php:
```diff
@@ -0,0 +1,36 @@
+ */
+final class PhoneNumberFactory
+{
+    use ConfigurationTrait;
```
Similar discussion as above around trait replacement

Ocramius commented on src/View/Helper/Factory/PhoneNumberFormatFactory.php:
```diff
@@ -0,0 +1,19 @@
+
+final class PhoneNumberFormatFactory
+{
+    use ConfigurationTrait;
```
Similar discussion as above around trait replacement

Ocramius commented on src/View/Helper/PhoneNumberFormat.php:
```diff
@@ -0,0 +1,94 @@
+     * @param non-empty-string      $number
+     * @param non-empty-string|null $countryCodeOrLocale
+     */
+    public function toPhoneNumber(string $number, ?string $countryCodeOrLocale = null): ?PhoneNumberValue
```
Should this be `tryToPhoneNumber()`?

Ocramius commented on src/View/Helper/PhoneNumberFormat.php:
```diff
@@ -0,0 +1,94 @@
+    {
+        $countryCodeOrLocale = $countryCodeOrLocale ? CountryCode::tryFromString($countryCodeOrLocale) : null;
+
+        return $countryCodeOrLocale ?: $this->country;
```
If the values are always `CountryCode|null`, let's use `??` instead of `?:` (which does implicitly cast)

Ocramius reviewed (changes requested):
It took months, but I finally had some space in my head to look at this.

I did a first skim at code level: implementation details.

I don't have an architectural overview yet - may need another pass for that to be clear to me.

Ocramius commented:
I renamed `main` to `1.0.x`

gsteel committed:
Fix trailing slash in autoload and update lock file including latest dependencies

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Drop PHP 7.4 and 8.0

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Improve CountryCode value object

- Use PHP 8.1 features making the code public readonly
- Extract exception messages into static constructors in the exception class
- Add method documentation
- Mark appropriate methods as pure

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Improve Module and ConfigProvider

- Add explanatory notes about the purpose of the classes
- improve type inference for consumers

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Change configuration trait to an abstract class with public methods

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Rename test namespace so that we can use psalm-internal classes within tests

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Marks all static exception constructors as psalm-pure and remove evil empty()

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
CS Fix

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Replace conditional with null coalesce, removing function call

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Improve abstract filter

- tighten up inheritance
- make use of PHP 8.1 features
- rename phone number factory method
- mark as internal

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Improve validator

- Instead of forcing the Options type, check and assert its shape
- Remove evil empty() call
- Don't assign ignored exceptions to variables

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Remove redundant exception assignments

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Improve phone number value object

- use constructor property promotion
- mark the country code and short number flag as readonly public properties
- remove the getters for the now public properties
- remove empty()
- document named constructor

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Improve formatting view helper

- Use constructor property promotion
- Use null coalesce where appropriate
- rename `toPhoneNumber` to `tryToPhoneNumber` to improve semantics

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Rename test case to match subject

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Make the `tryToPhoneNumber()` method of the view helper private and remove associated tests

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel commented:
Thanks @Ocramius I'd appreciate some more feedback on a couple of outstanding things 👍

froschdesign commented on src/CountryCode.php:
```diff
@@ -0,0 +1,141 @@
+/**
+ * @psalm-immutable
+ */
+final class CountryCode
```
Should this rather be placed in laminas-i18n? I think the goal should be the same handling for all i18n components. This means that in the future we will need the value object also in laminas-i18n.

gsteel replied:
At the moment, `i18n` is not a dependency so moving it there would make it one. Also, it's got some `8.1` specific syntax which isn't a drama to downgrade, but that'd be additional discussion - i.e. we could also bump `i18n` to `8.1` minimum (?) - cons aside, the value object would surely be useful and I think you're right that something this basic would be better placed in `i18n`

froschdesign replied:
> we could also bump `i18n` to `8.1` minimum (?)

Or 8.0? (Version 7 is nearly dead.)

gsteel replied:
Happy to send in a PR to `i18n` for this @froschdesign - Shall I just do it, or do you think we need more discussion on the decision?

froschdesign replied:
Version 7.4 has already been dropped in other packages, so there is no need for a deeper discussion on this topic.
Moving the value object to hold and detect a locale is also needed in laminas-i18n or will be used there in the future. It is easier to use.
And I don't see it as a problem that a dependency is created, because I would have expected it anyway. 😄

gsteel replied:
I can make the relevant changes here if this is merged: https://github.com/laminas/laminas-i18n/pull/84

gsteel committed:
Depend on laminas-i18n for CountryCode value object

Signed-off-by: George Steel <george@net-glue.co.uk>

internalsystemerror commented on src/Filter/Factory/ToE164Factory.php:
```diff
@@ -0,0 +1,17 @@
+{
+    public function __invoke(ContainerInterface $container): ToE164
+    {
+        return new ToE164(Configuration::defaultCountryCode($container));
```
Is the intention that these filters will only be used one at a time? Why not have the default country code as a service retrievable from the container? That could then be used for each of these different filters.

gsteel replied:
I think that leaving this as a configuration option is the best plan - if I configure `Laminas\i18n\CountryCode::class` to return the default country as configured specifically for the phone number lib, and then `Laminas\i18n` configures a service with the same name, the meaning of the object may well be at crossed purposes, not to mention that once implemented, it'll be a BC break to remove it. The way it works right now with a static method is un-ambiguous and there's no risk of the setup colliding with another package. We could set up a marker interface such as `Laminas\i18n\PhoneNumber\DefaultCountryCode` but we would forever have to `/** @psalm-var CountryCode $code **/` the thing we fetch from the container which imo is more work than it's worth.

internalsystemerror commented on src/Filter/Factory/ToInternationalPhoneNumberFactory.php:
```diff
@@ -0,0 +1,17 @@
+{
+    public function __invoke(ContainerInterface $container): ToInternationalPhoneNumber
+    {
+        return new ToInternationalPhoneNumber(Configuration::defaultCountryCode($container));
```
Is the intention that these filters will only be used one at a time? Why not have the default country code as a service retrievable from the container? That could then be used for each of these different filters.

internalsystemerror commented on src/Filter/Factory/ToNationalPhoneNumberFactory.php:
```diff
@@ -0,0 +1,17 @@
+{
+    public function __invoke(ContainerInterface $container): ToNationalPhoneNumber
+    {
+        return new ToNationalPhoneNumber(Configuration::defaultCountryCode($container));
```
Is the intention that these filters will only be used one at a time? Why not have the default country code as a service retrievable from the container? That could then be used for each of these different filters.

internalsystemerror commented on src/Filter/AbstractFilter.php:
```diff
@@ -0,0 +1,43 @@
+            return $value;
+        }
+
+        if (! is_scalar($value)) {
```
Is it worth adding a check for `Stringable` so that objects that can be cast to a string will be?

internalsystemerror commented on src/PhoneNumberValue.php:
```diff
@@ -0,0 +1,207 @@
+
+        $isShortNumber = $short->isValidShortNumberForRegion($prototype, $regionCode)
+            || $short->isValidShortNumber($prototype);
+        $isValidNumber = $util->isValidNumberForRegion($prototype, $regionCode)
```
This could be considered true unless `$isShortNumber` is false?

```php
$isValidNumber = $isShortNumber ? true : (
    $util->isValidNumberForRegion($prototype, $regionCode)
    || $util->isValidNumber($prototype)
);
```

gsteel replied:
👍 The exception case is more understandable when reduced to a single variable

internalsystemerror commented on src/PhoneNumberValue.php:
```diff
@@ -0,0 +1,207 @@
+        $isValidNumber = $util->isValidNumberForRegion($prototype, $regionCode)
+            || $util->isValidNumber($prototype);
+
+        if (! $isShortNumber && ! $isValidNumber) {
```
And then only `!$isValidNumber` needs to be checked here (assuming the above)

internalsystemerror commented on src/PhoneNumberValue.php:
```diff
@@ -0,0 +1,207 @@
+
+    public function toE164(): string
+    {
+        return PhoneNumberUtil::getInstance()->format($this->number, PhoneNumberFormat::E164);
```
Why not store the instance from `getInstance()` in constructor if it's retrieved and used often?

gsteel replied:
The singleton retrieved here is "big" - adding it as a property would _probably_ mean serialisation problems. `PhoneNumberValue` itself is not currently serialisable or `JsonSerializable` but it's feasible that this behaviour might be useful in future

internalsystemerror commented on src/Validator/PhoneNumber.php:
```diff
@@ -0,0 +1,142 @@
+    /** @param mixed $value */
+    public function isValid($value): bool
+    {
+        if (! is_scalar($value)) {
```
Again RE `Stringable` objects here, should they be checked for?

internalsystemerror commented on src/View/Helper/Factory/PhoneNumberFormatFactory.php:
```diff
@@ -0,0 +1,17 @@
+{
+    public function __invoke(ContainerInterface $container): PhoneNumberFormat
+    {
+        return new PhoneNumberFormat(Configuration::defaultCountryCode($container));
```
Another instance where the default Country Code could be reused.

internalsystemerror reviewed (commented):
I've added some minor nitpicks which you can accept/reject at your leisure

gsteel committed:
Modify filters and `PhoneNumber` validator to allow objects that implement stringable

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Reduce exception case to single variable

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Update locked deps, bump dev deps, cs fixes

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel commented:
Thanks for the reviews @internalsystemerror , @Ocramius and @froschdesign - can this PR be pushed over the finish line?

internalsystemerror reviewed (approved):
Looks good to me, and to merge

weierophinney commented on src/Exception/InvalidPhoneNumberException.php:
```diff
@@ -0,0 +1,44 @@
+final class InvalidPhoneNumberException extends InvalidArgumentException implements ExceptionInterface
+{
+    /** @psalm-pure */
+    public static function with(string $phoneNumber, ?string $countryCode): self
```
I'm wondering if this should be split into 3 separate types:

- InvalidPhoneNumberException
- InvalidPhoneNumberForRegionException
- UndetectablePhoneNumberRegionException

These could have a shared marker interface they all implement, if there's a desire to catch more generically.

I'm finding that when an exception can be used for multiple distinct error types, during testing, we have to validate that against expected messages, and that becomes problematic when the message changes. Having discrete exception types simplifies the testing, gives more granularity when catching, and, if a marker interface is used to group exceptions by "type", you can catch more generically if desired.

More classes, but better flexibility.

gsteel replied:
Thanks @weierophinney - this also revealed some dead code for removal too 👍

weierophinney commented on src/Factory/Configuration.php:
```diff
@@ -0,0 +1,39 @@
+ * @psalm-import-type PackageConfig from ConfigProvider
+ * @psalm-internal \Laminas\I18n\PhoneNumber
+ */
+abstract class Configuration // phpcs:ignore
```
From what I can see, you're consuming the static methods from this class, but never extending it. Would it make more sense to make this class final instead?

internalsystemerror replied:
with a private constructor if its not supposed to be called

gsteel replied:
Yes - it's effectively `Utils`!! - updated to final with private constructor. Already marked as internal.

weierophinney reviewed (commented):
Looks really good! Just have a couple of questions and suggestions.

gsteel committed:
Mark the configuration class as final with a private constructor rather than abstract

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Re-work exception classes with more specific types rather than varying messages

Signed-off-by: George Steel <george@net-glue.co.uk>

weierophinney reviewed (approved):
Looks great! :ship:

gsteel commented:
Thanks everyone for the reviews, help and suggestions!
